controllers/config/k3osconfig_controller.go

```diff
@@ -22,151 +22,163 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-	ctx := context.Background()
-
-	config, response, err := r.fetchK3OSConfig(ctx, req.NamespacedName)
-	if err != nil {
+func (r *K3OSConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
+	config := &configv1alpha1.K3OSConfig{}
+	if err := r.client.Get(ctx, req.NamespacedName, config); err != nil {
+		if apierrors.IsNotFound(err) { // request object not found, could have been deleted after reconcile request, return and don't requeue
+			return result, nil
+		}
 		r.logger.Error(err, "failed to fetch K3OSConfig")
-		return response.result, response.err
+		return result, err
 	}
-
+	config = config.DeepCopy()
 	r.logger.Info("successfully fetched K3OSConfig", "spec", config.Spec)
 
 	switch r.leader {
 	case true: // this instance of the operator won the leader election and can update the K3OSConfig CR
-		return r.handleK3OSConfigAsLeader(ctx, config)
+		result, err = r.handleK3OSConfigAsLeader(ctx, config)
 	default: // handle k3os config file
-		return r.handleK3OSConfig(ctx, config)
+		result, err = r.handleK3OSConfig(ctx, config)
 	}
-}
-
-func (r *K3OSConfigReconciler) handleK3OSConfigAsLeader(ctx context.Context, config *configv1alpha1.K3OSConfig) (ctrl.Result, error) {
-	return r.defaultRequeueResponse, nil
-}
-
-func (r *K3OSConfigReconciler) handleK3OSConfig(ctx context.Context, config *configv1alpha1.K3OSConfig) (ctrl.Result, error) {
-	return r.defaultRequeueResponse, nil
-}
 
-func (r *K3OSConfigReconciler) fetchK3OSConfig(ctx context.Context, name types.NamespacedName) (*configv1alpha1.K3OSConfig, *response, error) {
-	config := &configv1alpha1.K3OSConfig{}
-	if err := r.client.Get(ctx, name, config); err != nil {
-		if errors.IsNotFound(err) { // request object not found, could have been deleted after reconcile request, return and don't requeue
-			return nil, &response{result: ctrl.Result{}, err: nil}, err
-		}
-		return nil, &response{result: r.defaultRequeueResponse, err: nil}, err
+	if err != nil {
+		r.logger.Error(err, "failed to handle K3OSConfig")
 	}
-	return config.DeepCopy(), nil, nil
+	return result, err
 }
 
-// Option denotes an option for configuring this controller.
-type Option interface{}
-
-type requireLeaderElectionOpt struct {
-	requireLeaderElection bool
+func (r *K3OSConfigReconciler) handleK3OSConfigAsLeader(ctx context.Context, config *configv1alpha1.K3OSConfig) (ctrl.Result, error) {
+	return ctrl.Result{}, nil
 }
 
-// RequireLeaderElection returns an option that requires the operator being the leader to run this controller instance.
-func RequireLeaderElection() Option {
-	return &requireLeaderElectionOpt{}
+func resultError(err error) error {
+	switch {
+	case err == nil:
+		return nil
+	case errors.Is(err, errors.ErrSkipUpdate):
+		return nil
+	case apierrors.IsNotFound(err), apierrors.IsGone(err):
+		// log error
+		return nil
+	case apierrors.IsForbidden(err):
+		// log error
```

Comment:
Add the log statement – pass the logger into this function for that.

Also, the log message should include something along the lines of: `Did you forget to apply some RBAC rules?`.

Reply:
done